// next/dist/pages/_app.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var utils_1 = require("next-server/dist/lib/utils");
var router_1 = require("next/router");
var App = /** @class */ (function (_super) {
    __extends(App, _super);
    function App() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    App.getInitialProps = function (_a) {
        var Component = _a.Component, router = _a.router, ctx = _a.ctx;
        return __awaiter(this, void 0, void 0, function () {
            var pageProps;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, utils_1.loadGetInitialProps(Component, ctx)];
                    case 1:
                        pageProps = _b.sent();
                        return [2 /*return*/, { pageProps: pageProps }];
                }
            });
        });
    };
    App.prototype.getChildContext = function () {
        return {
            router: router_1.makePublicRouterInstance(this.props.router)
        };
    };
    // Kept here for backwards compatibility.
    // When someone ended App they could call `super.componentDidCatch`. This is now deprecated.
    App.prototype.componentDidCatch = function (err) {
        throw err;
    };
    App.prototype.render = function () {
        var _a = this.props, router = _a.router, Component = _a.Component, pageProps = _a.pageProps;
        var url = createUrl(router);
        return (react_1.default.createElement(Container, null,
            react_1.default.createElement(Component, __assign({}, pageProps, { url: url }))));
    };
    App.childContextTypes = {
        router: prop_types_1.default.object
    };
    return App;
}(react_1.Component));
exports.default = App;
var Container = /** @class */ (function (_super) {
    __extends(Container, _super);
    function Container() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Container.prototype.componentDidMount = function () {
        this.scrollToHash();
    };
    Container.prototype.componentDidUpdate = function () {
        this.scrollToHash();
    };
    Container.prototype.scrollToHash = function () {
        var hash = window.location.hash;
        hash = hash ? hash.substring(1) : false;
        if (!hash)
            return;
        var el = document.getElementById(hash);
        if (!el)
            return;
        // If we call scrollIntoView() in here without a setTimeout
        // it won't scroll properly.
        setTimeout(function () { return el.scrollIntoView(); }, 0);
    };
    Container.prototype.render = function () {
        return this.props.children;
    };
    return Container;
}(react_1.Component));
exports.Container = Container;
var warnUrl = utils_1.execOnce(function () {
    if (process.env.NODE_ENV !== 'production') {
        console.error("Warning: the 'url' property is deprecated. https://err.sh/zeit/next.js/url-deprecated");
    }
});
function createUrl(router) {
    // This is to make sure we don't references the router object at call time
    var pathname = router.pathname, asPath = router.asPath, query = router.query;
    return {
        get query() {
            warnUrl();
            return query;
        },
        get pathname() {
            warnUrl();
            return pathname;
        },
        get asPath() {
            warnUrl();
            return asPath;
        },
        back: function () {
            warnUrl();
            router.back();
        },
        push: function (url, as) {
            warnUrl();
            return router.push(url, as);
        },
        pushTo: function (href, as) {
            warnUrl();
            var pushRoute = as ? href : null;
            var pushUrl = as || href;
            return router.push(pushRoute, pushUrl);
        },
        replace: function (url, as) {
            warnUrl();
            return router.replace(url, as);
        },
        replaceTo: function (href, as) {
            warnUrl();
            var replaceRoute = as ? href : null;
            var replaceUrl = as || href;
            return router.replace(replaceRoute, replaceUrl);
        }
    };
}
exports.createUrl = createUrl;
