// next/dist/pages/_document.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable */
var react_1 = __importStar(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var htmlescape_1 = __importDefault(require("htmlescape"));
var server_1 = __importDefault(require("styled-jsx/server"));
var Fragment = react_1.default.Fragment || function Fragment(_a) {
    var children = _a.children;
    return react_1.default.createElement("div", null, children);
};
var Document = /** @class */ (function (_super) {
    __extends(Document, _super);
    function Document() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Document.getInitialProps = function (_a) {
        var renderPage = _a.renderPage;
        var _b = renderPage(), html = _b.html, head = _b.head;
        var styles = server_1.default();
        return { html: html, head: head, styles: styles };
    };
    Document.prototype.getChildContext = function () {
        return { _documentProps: this.props };
    };
    Document.prototype.render = function () {
        return react_1.default.createElement("html", null,
            react_1.default.createElement(Head, null),
            react_1.default.createElement("body", null,
                react_1.default.createElement(Main, null),
                react_1.default.createElement(NextScript, null)));
    };
    Document.childContextTypes = {
        _documentProps: prop_types_1.default.any
    };
    return Document;
}(react_1.Component));
exports.default = Document;
var Head = /** @class */ (function (_super) {
    __extends(Head, _super);
    function Head() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Head.prototype.getCssLinks = function () {
        var _this = this;
        var _a = this.context._documentProps, assetPrefix = _a.assetPrefix, files = _a.files;
        if (!files || files.length === 0) {
            return null;
        }
        return files.map(function (file) {
            // Only render .css files here
            if (!/\.css$/.exec(file)) {
                return null;
            }
            return react_1.default.createElement("link", { key: file, nonce: _this.props.nonce, rel: 'stylesheet', href: assetPrefix + "/_next/" + file, crossOrigin: _this.props.crossOrigin || process.crossOrigin });
        });
    };
    Head.prototype.getPreloadDynamicChunks = function () {
        var _this = this;
        var _a = this.context._documentProps, dynamicImports = _a.dynamicImports, assetPrefix = _a.assetPrefix;
        return dynamicImports.map(function (bundle) {
            return react_1.default.createElement("link", { rel: 'preload', key: bundle.file, href: assetPrefix + "/_next/" + bundle.file, as: 'script', nonce: _this.props.nonce, crossOrigin: _this.props.crossOrigin || process.crossOrigin });
        });
    };
    Head.prototype.getPreloadMainLinks = function () {
        var _this = this;
        var _a = this.context._documentProps, assetPrefix = _a.assetPrefix, files = _a.files;
        if (!files || files.length === 0) {
            return null;
        }
        return files.map(function (file) {
            // Only render .js files here
            if (!/\.js$/.exec(file)) {
                return null;
            }
            return react_1.default.createElement("link", { key: file, nonce: _this.props.nonce, rel: 'preload', href: assetPrefix + "/_next/" + file, as: 'script', crossOrigin: _this.props.crossOrigin || process.crossOrigin });
        });
    };
    Head.prototype.render = function () {
        var _a = this.context._documentProps, head = _a.head, styles = _a.styles, assetPrefix = _a.assetPrefix, __NEXT_DATA__ = _a.__NEXT_DATA__;
        var page = __NEXT_DATA__.page, buildId = __NEXT_DATA__.buildId;
        var pagePathname = getPagePathname(page);
        var children = this.props.children;
        // show a warning if Head contains <title> (only in development)
        if (process.env.NODE_ENV !== 'production') {
            children = react_1.default.Children.map(children, function (child) {
                if (child && child.type === 'title') {
                    console.warn("Warning: <title> should not be used in _document.js's <Head>. https://err.sh/next.js/no-document-title");
                }
                return child;
            });
            if (this.props.crossOrigin)
                console.warn('Warning: `Head` attribute `crossOrigin` is deprecated. https://err.sh/next.js/doc-crossorigin-deprecated');
        }
        return react_1.default.createElement("head", __assign({}, this.props),
            children,
            head,
            page !== '/_error' && react_1.default.createElement("link", { rel: 'preload', href: assetPrefix + "/_next/static/" + buildId + "/pages" + pagePathname, as: 'script', nonce: this.props.nonce, crossOrigin: this.props.crossOrigin || process.crossOrigin }),
            react_1.default.createElement("link", { rel: 'preload', href: assetPrefix + "/_next/static/" + buildId + "/pages/_app.js", as: 'script', nonce: this.props.nonce, crossOrigin: this.props.crossOrigin || process.crossOrigin }),
            react_1.default.createElement("link", { rel: 'preload', href: assetPrefix + "/_next/static/" + buildId + "/pages/_error.js", as: 'script', nonce: this.props.nonce, crossOrigin: this.props.crossOrigin || process.crossOrigin }),
            this.getPreloadDynamicChunks(),
            this.getPreloadMainLinks(),
            this.getCssLinks(),
            styles || null);
    };
    Head.contextTypes = {
        _documentProps: prop_types_1.default.any
    };
    Head.propTypes = {
        nonce: prop_types_1.default.string,
        crossOrigin: prop_types_1.default.string
    };
    return Head;
}(react_1.Component));
exports.Head = Head;
var Main = /** @class */ (function (_super) {
    __extends(Main, _super);
    function Main() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Main.prototype.render = function () {
        var html = this.context._documentProps.html;
        return (react_1.default.createElement("div", { id: '__next', dangerouslySetInnerHTML: { __html: html } }));
    };
    Main.contextTypes = {
        _documentProps: prop_types_1.default.any
    };
    return Main;
}(react_1.Component));
exports.Main = Main;
var NextScript = /** @class */ (function (_super) {
    __extends(NextScript, _super);
    function NextScript() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NextScript.prototype.getDynamicChunks = function () {
        var _this = this;
        var _a = this.context._documentProps, dynamicImports = _a.dynamicImports, assetPrefix = _a.assetPrefix;
        return dynamicImports.map(function (bundle) {
            return react_1.default.createElement("script", { async: true, key: bundle.file, src: assetPrefix + "/_next/" + bundle.file, nonce: _this.props.nonce, crossOrigin: _this.props.crossOrigin || process.crossOrigin });
        });
    };
    NextScript.prototype.getScripts = function () {
        var _this = this;
        var _a = this.context._documentProps, assetPrefix = _a.assetPrefix, files = _a.files;
        if (!files || files.length === 0) {
            return null;
        }
        return files.map(function (file) {
            // Only render .js files here
            if (!/\.js$/.exec(file)) {
                return null;
            }
            return react_1.default.createElement("script", { key: file, src: assetPrefix + "/_next/" + file, nonce: _this.props.nonce, async: true, crossOrigin: _this.props.crossOrigin || process.crossOrigin });
        });
    };
    NextScript.getInlineScriptSource = function (documentProps) {
        var __NEXT_DATA__ = documentProps.__NEXT_DATA__;
        return htmlescape_1.default(__NEXT_DATA__);
    };
    NextScript.prototype.render = function () {
        var _this = this;
        var _a = this.context._documentProps, staticMarkup = _a.staticMarkup, assetPrefix = _a.assetPrefix, devFiles = _a.devFiles, __NEXT_DATA__ = _a.__NEXT_DATA__;
        var page = __NEXT_DATA__.page, buildId = __NEXT_DATA__.buildId;
        var pagePathname = getPagePathname(page);
        if (process.env.NODE_ENV !== 'production') {
            if (this.props.crossOrigin)
                console.warn('Warning: `NextScript` attribute `crossOrigin` is deprecated. https://err.sh/next.js/doc-crossorigin-deprecated');
        }
        return react_1.default.createElement(Fragment, null,
            devFiles ? devFiles.map(function (file) { return react_1.default.createElement("script", { key: file, src: assetPrefix + "/_next/" + file, nonce: _this.props.nonce, crossOrigin: _this.props.crossOrigin || process.crossOrigin }); }) : null,
            staticMarkup ? null : react_1.default.createElement("script", { id: "__NEXT_DATA__", type: "application/json", nonce: this.props.nonce, crossOrigin: this.props.crossOrigin || process.crossOrigin, dangerouslySetInnerHTML: {
                    __html: NextScript.getInlineScriptSource(this.context._documentProps)
                } }),
            page !== '/_error' && react_1.default.createElement("script", { async: true, id: "__NEXT_PAGE__" + page, src: assetPrefix + "/_next/static/" + buildId + "/pages" + pagePathname, nonce: this.props.nonce, crossOrigin: this.props.crossOrigin || process.crossOrigin }),
            react_1.default.createElement("script", { async: true, id: "__NEXT_PAGE__/_app", src: assetPrefix + "/_next/static/" + buildId + "/pages/_app.js", nonce: this.props.nonce, crossOrigin: this.props.crossOrigin || process.crossOrigin }),
            react_1.default.createElement("script", { async: true, id: "__NEXT_PAGE__/_error", src: assetPrefix + "/_next/static/" + buildId + "/pages/_error.js", nonce: this.props.nonce, crossOrigin: this.props.crossOrigin || process.crossOrigin }),
            staticMarkup ? null : this.getDynamicChunks(),
            staticMarkup ? null : this.getScripts());
    };
    NextScript.contextTypes = {
        _documentProps: prop_types_1.default.any
    };
    NextScript.propTypes = {
        nonce: prop_types_1.default.string,
        crossOrigin: prop_types_1.default.string
    };
    return NextScript;
}(react_1.Component));
exports.NextScript = NextScript;
function getPagePathname(page) {
    if (page === '/') {
        return '/index.js';
    }
    return page + ".js";
}
