// next/dist/pages/_error.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var http_status_1 = __importDefault(require("http-status"));
var head_1 = __importDefault(require("next-server/head"));
var Error = /** @class */ (function (_super) {
    __extends(Error, _super);
    function Error() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Error.getInitialProps = function (_a) {
        var res = _a.res, err = _a.err;
        var statusCode = res ? res.statusCode : (err ? err.statusCode : null);
        return { statusCode: statusCode };
    };
    Error.prototype.render = function () {
        var statusCode = this.props.statusCode;
        var title = statusCode === 404
            ? 'This page could not be found'
            : http_status_1.default[statusCode] || 'An unexpected error has occurred';
        return react_1.default.createElement("div", { style: styles.error },
            react_1.default.createElement(head_1.default, null,
                react_1.default.createElement("meta", { name: 'viewport', content: 'width=device-width, initial-scale=1.0' }),
                react_1.default.createElement("title", null,
                    statusCode,
                    ": ",
                    title)),
            react_1.default.createElement("div", null,
                react_1.default.createElement("style", { dangerouslySetInnerHTML: { __html: 'body { margin: 0 }' } }),
                statusCode ? react_1.default.createElement("h1", { style: styles.h1 }, statusCode) : null,
                react_1.default.createElement("div", { style: styles.desc },
                    react_1.default.createElement("h2", { style: styles.h2 },
                        title,
                        "."))));
    };
    Error.displayName = 'ErrorPage';
    return Error;
}(react_1.default.Component));
exports.default = Error;
if (process.env.NODE_ENV !== 'production') {
    Error.propTypes = {
        statusCode: prop_types_1.default.number
    };
}
var styles = {
    error: {
        color: '#000',
        background: '#fff',
        fontFamily: '-apple-system, BlinkMacSystemFont, Roboto, "Segoe UI", "Fira Sans", Avenir, "Helvetica Neue", "Lucida Grande", sans-serif',
        height: '100vh',
        textAlign: 'center',
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center'
    },
    desc: {
        display: 'inline-block',
        textAlign: 'left',
        lineHeight: '49px',
        height: '49px',
        verticalAlign: 'middle'
    },
    h1: {
        display: 'inline-block',
        borderRight: '1px solid rgba(0, 0, 0,.3)',
        margin: 0,
        marginRight: '20px',
        padding: '10px 23px 10px 0',
        fontSize: '24px',
        fontWeight: 500,
        verticalAlign: 'top'
    },
    h2: {
        fontSize: '14px',
        fontWeight: 'normal',
        lineHeight: 'inherit',
        margin: 0,
        padding: 0
    }
};
