// semantic-ui-calendar-react/dist/views/DatesRangeView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _lodash = _interopRequireDefault(require("lodash"));

var _Calendar = _interopRequireDefault(require("./Calendar"));

var _Header = _interopRequireDefault(require("./CalendarHeader/Header"));

var _Body = _interopRequireDefault(require("./CalendarBody/Body"));

var _DayView = require("./DayView");

var _BaseView2 = _interopRequireDefault(require("./BaseView"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DAY_POSITIONS = _lodash.default.range(_DayView.WEEKS_TO_DISPLAY * 7);

function getActive(start, end) {
  if (_lodash.default.isNil(start) && _lodash.default.isNil(end)) return;

  if (!_lodash.default.isNil(start) && _lodash.default.isNil(end)) {
    return start;
  }

  if (!_lodash.default.isNil(start) && !_lodash.default.isNil(end)) {
    return DAY_POSITIONS.slice(start, end + 1);
  }
}

var DatesRangeView =
/*#__PURE__*/
function (_BaseView) {
  _inherits(DatesRangeView, _BaseView);

  function DatesRangeView() {
    _classCallCheck(this, DatesRangeView);

    return _possibleConstructorReturn(this, _getPrototypeOf(DatesRangeView).apply(this, arguments));
  }

  _createClass(DatesRangeView, [{
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props = this.props,
          days = _this$props.days,
          onNextPageBtnClick = _this$props.onNextPageBtnClick,
          onPrevPageBtnClick = _this$props.onPrevPageBtnClick,
          onDayClick = _this$props.onDayClick,
          hasPrevPage = _this$props.hasPrevPage,
          hasNextPage = _this$props.hasNextPage,
          currentDate = _this$props.currentDate,
          onHeaderClick = _this$props.onHeaderClick,
          active = _this$props.active,
          disabled = _this$props.disabled,
          selectedRange = _this$props.selectedRange,
          hovered = _this$props.hovered,
          onCellHover = _this$props.onCellHover,
          onMount = _this$props.onMount,
          rest = _objectWithoutProperties(_this$props, ["days", "onNextPageBtnClick", "onPrevPageBtnClick", "onDayClick", "hasPrevPage", "hasNextPage", "currentDate", "onHeaderClick", "active", "disabled", "selectedRange", "hovered", "onCellHover", "onMount"]);

      var start = active.start,
          end = active.end;
      return _react.default.createElement(_Calendar.default, _extends({
        ref: function ref(e) {
          return _this.calendarNode = _reactDom.default.findDOMNode(e);
        }
      }, rest), _react.default.createElement(_Header.default, {
        width: _DayView.DAY_CALENDAR_ROW_WIDTH,
        displayWeeks: true,
        rangeRowContent: selectedRange,
        onNextPageBtnClick: onNextPageBtnClick,
        onPrevPageBtnClick: onPrevPageBtnClick,
        hasNextPage: hasNextPage,
        hasPrevPage: hasPrevPage,
        title: currentDate,
        onHeaderClick: onHeaderClick
      }), _react.default.createElement(_Body.default, {
        width: _DayView.DAY_CALENDAR_ROW_WIDTH,
        data: days,
        onCellClick: onDayClick,
        onCellHover: onCellHover,
        hovered: hovered,
        active: getActive(start, end),
        disabled: disabled
      }));
    }
  }]);

  return DatesRangeView;
}(_BaseView2.default);

_defineProperty(DatesRangeView, "handledProps", ["active", "currentDate", "days", "disabled", "hasNextPage", "hasPrevPage", "hovered", "onCellHover", "onDayClick", "onHeaderClick", "onMount", "onNextPageBtnClick", "onPrevPageBtnClick", "selectedRange"]);

DatesRangeView.propTypes = {
  /** An array of dates to fill a calendar with. */
  days: _propTypes.default.arrayOf(_propTypes.default.string).isRequired,

  /** Called after click on next page button. */
  onNextPageBtnClick: _propTypes.default.func.isRequired,

  /** Called after click on previous page button. */
  onPrevPageBtnClick: _propTypes.default.func.isRequired,

  /** Called after click on day. */
  onDayClick: _propTypes.default.func.isRequired,

  /** Whether to display previous page button as active or disabled. */
  hasPrevPage: _propTypes.default.bool.isRequired,

  /** Whether to display next page button as active or disabled. */
  hasNextPage: _propTypes.default.bool.isRequired,

  /** Date that is displayed in calendar header. */
  currentDate: _propTypes.default.string.isRequired,

  /** Selected range that is displayed in calendar header. */
  selectedRange: _propTypes.default.string.isRequired,

  /** Start and end of a range of day positions to display as active. */
  active: _propTypes.default.shape({
    start: _propTypes.default.number,
    end: _propTypes.default.number
  }).isRequired,

  /** Called on calendar cell hover. */
  onCellHover: _propTypes.default.func,

  /** Index of a month that should be displayed as hovered. */
  hovered: _propTypes.default.number,

  /** Called after click on calendar header. */
  onHeaderClick: _propTypes.default.func,

  /** An array of day positions to display as disabled. */
  disabled: _propTypes.default.arrayOf(_propTypes.default.number),
  onMount: _propTypes.default.func
};
DatesRangeView.defaultProps = {
  active: {
    start: undefined,
    end: undefined
  }
};
var _default = DatesRangeView;
exports.default = _default;
