// semantic-ui-calendar-react/dist/views/MonthView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Calendar = _interopRequireDefault(require("./Calendar"));

var _Header = _interopRequireDefault(require("./CalendarHeader/Header"));

var _Body = _interopRequireDefault(require("./CalendarBody/Body"));

var _BaseView2 = _interopRequireDefault(require("./BaseView"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var MONTH_CALENDAR_ROW_WIDTH = '3';

var MonthView =
/*#__PURE__*/
function (_BaseView) {
  _inherits(MonthView, _BaseView);

  function MonthView() {
    _classCallCheck(this, MonthView);

    return _possibleConstructorReturn(this, _getPrototypeOf(MonthView).apply(this, arguments));
  }

  _createClass(MonthView, [{
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props = this.props,
          months = _this$props.months,
          hasHeader = _this$props.hasHeader,
          onMonthClick = _this$props.onMonthClick,
          onNextPageBtnClick = _this$props.onNextPageBtnClick,
          onPrevPageBtnClick = _this$props.onPrevPageBtnClick,
          hasPrevPage = _this$props.hasPrevPage,
          hasNextPage = _this$props.hasNextPage,
          onHeaderClick = _this$props.onHeaderClick,
          disabled = _this$props.disabled,
          active = _this$props.active,
          currentYear = _this$props.currentYear,
          onCellHover = _this$props.onCellHover,
          hovered = _this$props.hovered,
          onMount = _this$props.onMount,
          rest = _objectWithoutProperties(_this$props, ["months", "hasHeader", "onMonthClick", "onNextPageBtnClick", "onPrevPageBtnClick", "hasPrevPage", "hasNextPage", "onHeaderClick", "disabled", "active", "currentYear", "onCellHover", "hovered", "onMount"]);

      var headerProps = {
        onNextPageBtnClick: onNextPageBtnClick,
        onPrevPageBtnClick: onPrevPageBtnClick,
        hasPrevPage: hasPrevPage,
        hasNextPage: hasNextPage,
        onHeaderClick: onHeaderClick,
        title: currentYear,
        displayWeeks: false,
        width: MONTH_CALENDAR_ROW_WIDTH
      };
      return _react.default.createElement(_Calendar.default, _extends({
        ref: function ref(e) {
          return _this.calendarNode = _reactDom.default.findDOMNode(e);
        }
      }, rest), hasHeader && _react.default.createElement(_Header.default, headerProps), _react.default.createElement(_Body.default, {
        width: MONTH_CALENDAR_ROW_WIDTH,
        data: months,
        onCellClick: onMonthClick,
        onCellHover: onCellHover,
        active: active,
        hovered: hovered,
        disabled: disabled
      }));
    }
  }]);

  return MonthView;
}(_BaseView2.default);

_defineProperty(MonthView, "handledProps", ["active", "currentYear", "disabled", "hasHeader", "hasNextPage", "hasPrevPage", "hovered", "months", "onCellHover", "onHeaderClick", "onMonthClick", "onMount", "onNextPageBtnClick", "onPrevPageBtnClick"]);

MonthView.propTypes = {
  /** Array of months to fill a calendar with. */
  months: _propTypes.default.arrayOf(_propTypes.default.string).isRequired,

  /** Wether to display header or not. */
  hasHeader: _propTypes.default.bool.isRequired,

  /** Called after click on month. */
  onMonthClick: _propTypes.default.func.isRequired,

  /** Called on calendar cell hover. */
  onCellHover: _propTypes.default.func,

  /** Index of a month that should be displayed as hovered. */
  hovered: _propTypes.default.number,

  /** Called after click on next page button. */
  onNextPageBtnClick: _propTypes.default.func,

  /** Called after click on previous page button. */
  onPrevPageBtnClick: _propTypes.default.func,

  /** Whether to display previous page button as active or disabled. */
  hasPrevPage: _propTypes.default.bool,

  /** Whether to display next page button as active or disabled. */
  hasNextPage: _propTypes.default.bool,

  /** Called after click on calendar header. */
  onHeaderClick: _propTypes.default.func,

  /** An array of month indexes to display as disabled. */
  disabled: _propTypes.default.arrayOf(_propTypes.default.number),

  /** Index of a month that should be displayed as active. */
  active: _propTypes.default.number,

  /** A year to display in header. */
  currentYear: _propTypes.default.string,
  onMount: _propTypes.default.func
};
var _default = MonthView;
exports.default = _default;
