// semantic-ui-calendar-react/dist/views/YearView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _lodash = _interopRequireDefault(require("lodash"));

var _Calendar = _interopRequireDefault(require("./Calendar"));

var _Header = _interopRequireDefault(require("./CalendarHeader/Header"));

var _Body = _interopRequireDefault(require("./CalendarBody/Body"));

var _BaseView2 = _interopRequireDefault(require("./BaseView"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var YEAR_CALENDAR_ROW_WIDTH = '3';

var YearView =
/*#__PURE__*/
function (_BaseView) {
  _inherits(YearView, _BaseView);

  function YearView() {
    _classCallCheck(this, YearView);

    return _possibleConstructorReturn(this, _getPrototypeOf(YearView).apply(this, arguments));
  }

  _createClass(YearView, [{
    key: "render",
    value: function render() {
      var _this = this;

      var _this$props = this.props,
          years = _this$props.years,
          onNextPageBtnClick = _this$props.onNextPageBtnClick,
          onPrevPageBtnClick = _this$props.onPrevPageBtnClick,
          onYearClick = _this$props.onYearClick,
          hasNextPage = _this$props.hasNextPage,
          hasPrevPage = _this$props.hasPrevPage,
          onHeaderClick = _this$props.onHeaderClick,
          disabled = _this$props.disabled,
          active = _this$props.active,
          hovered = _this$props.hovered,
          onCellHover = _this$props.onCellHover,
          onMount = _this$props.onMount,
          rest = _objectWithoutProperties(_this$props, ["years", "onNextPageBtnClick", "onPrevPageBtnClick", "onYearClick", "hasNextPage", "hasPrevPage", "onHeaderClick", "disabled", "active", "hovered", "onCellHover", "onMount"]);

      var headerTitle = "".concat(_lodash.default.first(years), " - ").concat(_lodash.default.last(years));
      return _react.default.createElement(_Calendar.default, _extends({
        ref: function ref(e) {
          return _this.calendarNode = _reactDom.default.findDOMNode(e);
        }
      }, rest), _react.default.createElement(_Header.default, {
        title: headerTitle,
        onNextPageBtnClick: onNextPageBtnClick,
        onPrevPageBtnClick: onPrevPageBtnClick,
        hasNextPage: hasNextPage,
        hasPrevPage: hasPrevPage,
        onHeaderClick: onHeaderClick,
        width: YEAR_CALENDAR_ROW_WIDTH,
        displayWeeks: false
      }), _react.default.createElement(_Body.default, {
        width: YEAR_CALENDAR_ROW_WIDTH,
        data: years,
        hovered: hovered,
        onCellHover: onCellHover,
        onCellClick: onYearClick,
        active: active,
        disabled: disabled
      }));
    }
  }]);

  return YearView;
}(_BaseView2.default);

_defineProperty(YearView, "handledProps", ["active", "disabled", "hasNextPage", "hasPrevPage", "hovered", "onCellHover", "onHeaderClick", "onMount", "onNextPageBtnClick", "onPrevPageBtnClick", "onYearClick", "years"]);

YearView.propTypes = {
  /** An array of years to fill a calendar with. */
  years: _propTypes.default.arrayOf(_propTypes.default.string).isRequired,

  /** Called after click on next page button. */
  onNextPageBtnClick: _propTypes.default.func.isRequired,

  /** Called after click on previous page button. */
  onPrevPageBtnClick: _propTypes.default.func.isRequired,

  /** Called after click on year. */
  onYearClick: _propTypes.default.func.isRequired,

  /** Whether to display previous page button as active or disabled. */
  hasPrevPage: _propTypes.default.bool.isRequired,

  /** Whether to display next page button as active or disabled. */
  hasNextPage: _propTypes.default.bool.isRequired,

  /** Called on calendar cell hover. */
  onCellHover: _propTypes.default.func,

  /** Index of a cell that should be displayed as hovered. */
  hovered: _propTypes.default.number,

  /** Called after click on calendar header. */
  onHeaderClick: _propTypes.default.func,

  /** An array of numbers that represent indexes of years in `years` array that should be displayed as disabled. */
  disabled: _propTypes.default.arrayOf(_propTypes.default.number),

  /** Index of a year in `years` array that should be displayed as active. */
  active: _propTypes.default.number,
  onMount: _propTypes.default.func
};
var _default = YearView;
exports.default = _default;
