// next/dist/client/dev-error-overlay/hot-dev-client.js
/* eslint-disable camelcase */
/**
MIT License

Copyright (c) 2013-present, Facebook, Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
// This file is based on https://github.com/facebook/create-react-app/blob/v1.1.4/packages/react-dev-utils/webpackHotDevClient.js
// It's been edited to rely on webpack-hot-middleware and to be more compatible with SSR / Next.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var eventsource_1 = require("./eventsource");
var format_webpack_messages_1 = __importDefault(require("./format-webpack-messages"));
var ErrorOverlay = __importStar(require("react-error-overlay"));
var strip_ansi_1 = __importDefault(require("strip-ansi"));
var source_map_support_1 = require("../source-map-support");
var unfetch_1 = __importDefault(require("unfetch"));
// This alternative WebpackDevServer combines the functionality of:
// https://github.com/webpack/webpack-dev-server/blob/webpack-1/client/index.js
// https://github.com/webpack/webpack/blob/webpack-1/hot/dev-server.js
// It only supports their simplest configuration (hot updates on same server).
// It makes some opinionated choices on top, like adding a syntax error overlay
// that looks similar to our console output. The error overlay is inspired by:
// https://github.com/glenjamin/webpack-hot-middleware
// This is a modified version of create-react-app's webpackHotDevClient.js
// It implements webpack-hot-middleware's EventSource events instead of webpack-dev-server's websocket.
// https://github.com/facebook/create-react-app/blob/25184c4e91ebabd16fe1cde3d8630830e4a36a01/packages/react-dev-utils/webpackHotDevClient.js
var hadRuntimeError = false;
var customHmrEventHandler;
function connect(options) {
    // Open stack traces in an editor.
    ErrorOverlay.setEditorHandler(function editorHandler(_a) {
        var fileName = _a.fileName, lineNumber = _a.lineNumber, colNumber = _a.colNumber;
        // Resolve invalid paths coming from react-error-overlay
        var resolvedFilename = fileName.replace(/^webpack:\/\//, '');
        unfetch_1.default('/_next/development/open-stack-frame-in-editor' +
            ("?fileName=" + window.encodeURIComponent(resolvedFilename)) +
            ("&lineNumber=" + (lineNumber || 1)) +
            ("&colNumber=" + (colNumber || 1)));
    });
    // We need to keep track of if there has been a runtime error.
    // Essentially, we cannot guarantee application state was not corrupted by the
    // runtime error. To prevent confusing behavior, we forcibly reload the entire
    // application. This is handled below when we are notified of a compile (code
    // change).
    // See https://github.com/facebook/create-react-app/issues/3096
    ErrorOverlay.startReportingRuntimeErrors({
        onError: function () {
            hadRuntimeError = true;
        }
    });
    if (module.hot && typeof module.hot.dispose === 'function') {
        module.hot.dispose(function () {
            // TODO: why do we need this?
            ErrorOverlay.stopReportingRuntimeErrors();
        });
    }
    eventsource_1.getEventSourceWrapper(options).addMessageListener(function (event) {
        // This is the heartbeat event
        if (event.data === '\uD83D\uDC93') {
            return;
        }
        try {
            processMessage(event);
        }
        catch (ex) {
            console.warn('Invalid HMR message: ' + event.data + '\n' + ex);
        }
    });
    return {
        subscribeToHmrEvent: function (handler) {
            customHmrEventHandler = handler;
        },
        prepareError: function (err) {
            // Temporary workaround for https://github.com/facebook/create-react-app/issues/4760
            // Should be removed once the fix lands
            hadRuntimeError = true;
            // react-error-overlay expects a type of `Error`
            var error = new Error(err.message);
            error.name = err.name;
            error.stack = err.stack;
            // __NEXT_DIST_DIR is provided by webpack
            source_map_support_1.rewriteStacktrace(error, process.env.__NEXT_DIST_DIR);
            return error;
        }
    };
}
exports.default = connect;
// Remember some state related to hot module replacement.
var isFirstCompilation = true;
var mostRecentCompilationHash = null;
var hasCompileErrors = false;
function clearOutdatedErrors() {
    // Clean up outdated compile errors, if any.
    if (typeof console !== 'undefined' && typeof console.clear === 'function') {
        if (hasCompileErrors) {
            console.clear();
        }
    }
}
// Successful compilation.
function handleSuccess() {
    var isHotUpdate = !isFirstCompilation;
    isFirstCompilation = false;
    hasCompileErrors = false;
    // Attempt to apply hot updates or reload.
    if (isHotUpdate) {
        tryApplyUpdates(function onHotUpdateSuccess() {
            // Only dismiss it when we're sure it's a hot update.
            // Otherwise it would flicker right before the reload.
            ErrorOverlay.dismissBuildError();
        });
    }
}
// Compilation with warnings (e.g. ESLint).
function handleWarnings(warnings) {
    clearOutdatedErrors();
    // Print warnings to the console.
    var formatted = format_webpack_messages_1.default({
        warnings: warnings,
        errors: []
    });
    if (typeof console !== 'undefined' && typeof console.warn === 'function') {
        for (var i = 0; i < formatted.warnings.length; i++) {
            if (i === 5) {
                console.warn('There were more warnings in other files.\n' +
                    'You can find a complete log in the terminal.');
                break;
            }
            console.warn(strip_ansi_1.default(formatted.warnings[i]));
        }
    }
}
// Compilation with errors (e.g. syntax error or missing modules).
function handleErrors(errors) {
    clearOutdatedErrors();
    isFirstCompilation = false;
    hasCompileErrors = true;
    // "Massage" webpack messages.
    var formatted = format_webpack_messages_1.default({
        errors: errors,
        warnings: []
    });
    // Only show the first error.
    ErrorOverlay.reportBuildError(formatted.errors[0]);
    // Also log them to the console.
    if (typeof console !== 'undefined' && typeof console.error === 'function') {
        for (var i = 0; i < formatted.errors.length; i++) {
            console.error(strip_ansi_1.default(formatted.errors[i]));
        }
    }
}
// There is a newer version of the code available.
function handleAvailableHash(hash) {
    // Update last known compilation hash.
    mostRecentCompilationHash = hash;
}
// Handle messages from the server.
function processMessage(e) {
    var obj = JSON.parse(e.data);
    switch (obj.action) {
        case 'building': {
            console.log('[HMR] bundle ' + (obj.name ? "'" + obj.name + "' " : '') +
                'rebuilding');
            break;
        }
        case 'built':
        case 'sync': {
            clearOutdatedErrors();
            if (obj.hash) {
                handleAvailableHash(obj.hash);
            }
            if (obj.warnings.length > 0) {
                handleWarnings(obj.warnings);
                break;
            }
            if (obj.errors.length > 0) {
                // When there is a compilation error coming from SSR we have to reload the page on next successful compile
                if (obj.action === 'sync') {
                    hadRuntimeError = true;
                }
                handleErrors(obj.errors);
                break;
            }
            handleSuccess();
            break;
        }
        default: {
            if (customHmrEventHandler) {
                customHmrEventHandler(obj);
                break;
            }
            break;
        }
    }
}
// Is there a newer version of this code available?
function isUpdateAvailable() {
    /* globals __webpack_hash__ */
    // __webpack_hash__ is the hash of the current compilation.
    // It's a global variable injected by Webpack.
    return mostRecentCompilationHash !== __webpack_hash__;
}
// Webpack disallows updates in other states.
function canApplyUpdates() {
    return module.hot.status() === 'idle';
}
// Attempt to update code on the fly, fall back to a hard reload.
function tryApplyUpdates(onHotUpdateSuccess) {
    return __awaiter(this, void 0, void 0, function () {
        function handleApplyUpdates(err, updatedModules) {
            if (err || hadRuntimeError) {
                if (err) {
                    console.warn('Error while applying updates, reloading page', err);
                }
                if (hadRuntimeError) {
                    console.warn('Had runtime error previously, reloading page');
                }
                window.location.reload();
                return;
            }
            if (typeof onHotUpdateSuccess === 'function') {
                // Maybe we want to do something.
                onHotUpdateSuccess();
            }
            if (isUpdateAvailable()) {
                // While we were updating, there was a new update! Do it again.
                tryApplyUpdates();
            }
        }
        var updatedModules, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!module.hot) {
                        // HotModuleReplacementPlugin is not in Webpack configuration.
                        console.error('HotModuleReplacementPlugin is not in Webpack configuration.');
                        // window.location.reload();
                        return [2 /*return*/];
                    }
                    if (!isUpdateAvailable() || !canApplyUpdates()) {
                        return [2 /*return*/];
                    }
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, module.hot.check(/* autoApply */ {
                            ignoreUnaccepted: true
                        })];
                case 2:
                    updatedModules = _a.sent();
                    if (updatedModules) {
                        handleApplyUpdates(null, updatedModules);
                    }
                    return [3 /*break*/, 4];
                case 3:
                    err_1 = _a.sent();
                    handleApplyUpdates(err_1, null);
                    return [3 /*break*/, 4];
                case 4: return [2 /*return*/];
            }
        });
    });
}
